packages/protocol/contracts/Staking.sol

```diff
@@ -0,0 +1,213 @@
+        _deposit(_subject, _amount, _lockPeriod);
+    }
+
+    /**
+     * External function to buy & lock tokens.
+     * @param _subject Subject address for which tokens are getting deposited.
+     * @param _depositAmount amount of moxie tokens getting deposited.
+     */
+    function buyAndLock(address _subject, uint256 _depositAmount, uint256 _minReturnAmountAfterFee, uint256 _lockPeriod)
+        external
+        onlyValidLockPeriod(_lockPeriod)
+        nonReentrant
+    {
+        // transfer moxie to this contract
+        moxieToken.safeTransferFrom(msg.sender, address(this), _depositAmount);
+        // approve moxie bonding curve
+        moxieToken.approve(address(moxieBondingCurve), _depositAmount);
+        // pull moxie
+        uint256 _amount = moxieBondingCurve.buyShares(_subject, _depositAmount, _minReturnAmountAfterFee);
+        uint256 unlockTime = block.timestamp + _lockPeriod;
+        address subjectToken = address(tokenManager.tokens(_subject));
+        LockInfo memory lockInfo = LockInfo({
+            amount: _amount,
+            unlockTime: unlockTime,
+            subject: _subject,
+            subjectToken: subjectToken,
+            user: msg.sender,
+            lockPeriod: _lockPeriod
+        });
+        uint256 _index = lockCount++;
+        locks[_index] = lockInfo;
+        emit Lock(msg.sender, _subject, subjectToken, _index, _amount, unlockTime, _lockPeriod);
+    }
+
+    function withdraw(uint256[] memory _indexes, address _subject) external nonReentrant {
+        if (_indexes.length == 0) {
+            revert EmptyIndexes();
+        }
+        uint256 totalAmount = 0;
+        address _subjectToken;
+        for (uint256 i = 0; i < _indexes.length; i++) {
+            uint256 _index = _indexes[i];
+            LockInfo memory lockInfo = locks[_index];
+            if (_subject != lockInfo.subject) {
+                revert SubjectsDoesntMatch(_index);
+            }
+            if (lockInfo.unlockTime > block.timestamp) {
+                revert LockNotExpired(_index, block.timestamp, lockInfo.unlockTime);
+            }
+            if (lockInfo.user != msg.sender) {
+                revert NotOwner();
+            }
+            if (_subjectToken == address(0)) {
+                _subjectToken = lockInfo.subjectToken;
+            }
+            totalAmount += lockInfo.amount;
+            delete locks[_index];
+        }
+        IERC20Extended subjectToken = IERC20Extended(_subjectToken);
+        subjectToken.transfer(msg.sender, totalAmount);
+        emit Withdraw(msg.sender, _subject, _subjectToken, _indexes, totalAmount);
+    }
+
+    function extendLock(uint256[] memory _indexes, address _subject, uint256 _lockPeriod)
+        external
+        onlyValidLockPeriod(_lockPeriod)
+        nonReentrant
+    {
+        if (_indexes.length == 0) {
+            revert EmptyIndexes();
+        }
+        uint256 totalAmount = 0;
```

Comment:
extract a private function from withdraw & reuse it for whole logic except transfer.